plugins/webp-uploads/picture-element.php

```diff
@@ -0,0 +1,154 @@
+		}
+	);
+
+	// No eligible mime types.
+	if ( ! $mime_types ) {
+		return $image;
+	}
+
+	// If the original mime types is the only one available, no picture element is needed.
+	if ( 1 === count( $mime_types ) && $mime_types[0] === $original_file_mime_type ) {
+		return $image;
+	}
+
+	// Add each mime type to the picture's sources.
+	$picture_sources = '';
+
+	// Extract sizes using regex to parse image tag, then use to retrieve tag.
+	$image_sizes = array();
+	preg_match( '/width="([^"]+)"/', $image, $image_sizes );
+	$width = isset( $image_sizes[1] ) ? (int) $image_sizes[1] : false;
+	preg_match( '/height="([^"]+)"/', $image, $image_sizes );
+	$height      = isset( $image_sizes[1] ) ? (int) $image_sizes[1] : false;
+	$size_to_use = ( $width && $height ) ? array( $width, $height ) : 'full';
+
+	$image_src = wp_get_attachment_image_src( $attachment_id, $size_to_use, false );
+	if ( ! $image_src ) {
+		return $image;
+	}
+	list( $src, $width, $height ) = $image_src;
+	$size_array                   = array( absint( $width ), absint( $height ) );
+
+	foreach ( $mime_types as $image_mime_type ) {
+		$sizes = wp_calculate_image_sizes( $size_array, $src, $image_meta, $attachment_id );
+		// Filter core's wp_get_attachment_image_srcset to return the sources for the current mime type.
+
+		add_filter(
+			'wp_calculate_image_srcset',
+			static function ( $sources ) use ( $mime_type_data, $image_mime_type ) {
+				$filtered_sources = array();
+				foreach ( $sources as $source ) {
+					// Swap the URL for the current mime type.
+					if ( isset( $mime_type_data[ $image_mime_type ][ $source['descriptor'] ][ $source['value'] ] ) ) {
+						$filename  = $mime_type_data[ $image_mime_type ][ $source['descriptor'] ][ $source['value'] ]['file'];
+						$url_array = explode( '/', $source['url'] );
+						array_pop( $url_array );
+
+						$filtered_sources[] = array(
+							'url'        => implode( '/', $url_array ) . '/' . $filename,
+							'descriptor' => $source['descriptor'],
+							'value'      => $source['value'],
+						);
+					}
+				}
+				return $filtered_sources;
+			}
+		);
+		$image_srcset = wp_get_attachment_image_srcset( $attachment_id, $size_array, $image_meta );
+		remove_all_filters( 'wp_calculate_image_srcset' );
+
+		$picture_sources .= sprintf(
+			'<source type="%s" srcset="%s" sizes="%s">',
+			$image_mime_type,
+			$image_srcset,
+			$sizes
+		);
+	}
+
+	// Fall back to the original image without a srcset.
+	$original_sizes = array( $image_src[1], $image_src[2] );
+	$original_image = wp_get_original_image_url( $attachment_id );
+	// Fail gracefully if the original image is not found.
+	if ( ! $original_image ) {
+		return $image;
+	}
+	add_filter( 'wp_calculate_image_srcset_meta', '__return_false' );
+	$original_image_without_srcset = wp_get_attachment_image( $attachment_id, $original_sizes, false, array( 'src' => $original_image ) );
+	remove_filter( 'wp_calculate_image_srcset_meta', '__return_false' );
```

Comment:
Just in case another plugin has already added `__return_false` as a filter:

```suggestion
	$filter = function (): bool {
		return false;
	};
	add_filter( 'wp_calculate_image_srcset_meta', $filter );
	$original_image_without_srcset = wp_get_attachment_image( $attachment_id, $original_sizes, false, array( 'src' => $original_image ) );
	remove_filter( 'wp_calculate_image_srcset_meta', $filter );
```